### PR TITLE
Fix lookahead buffer size reported to littlefs2-sys

### PR DESCRIPTION
**Pull request by robin-nitrokey**
Previously, we reported the lookahead buffer size in bytes but littlefs2-sys expects the lookahead buffer size as a multiple of 8 bytes.  This could lead to a buffer overflow causing filesystem corruption.  This patch fixes the reported lookahead buffer size.

Note that Storage::LOOKAHEAD_WORDS_SIZE allows users to set invalid values (as it is measured in 4 bytes, not in 8 bytes).  Invalid values that were previously accepted because of the wrong buffer size calculation can now be rejected by littlefs2-sys.

This is a combination of two previous patches:
	https://github.com/trussed-dev/littlefs2/pull/19
	https://github.com/Nitrokey/littlefs2/pull/1

Fixes: https://github.com/trussed-dev/littlefs2/issues/16

----

I’m wondering whether we should treat this as a breaking change or not.  On the one hand, this is a bug fix.  On the other hand, it may cause code that previously worked (because the size calculation was wrong) to fail.  If we consider this to be a breaking change, we should also change `Storage::LOOKAHEAD_BUFFER_SIZE` so that it uses multiples of 8 and it is no longer possible to set invalid values.

Also, it would be good to have a test case that demonstrates the bug.

cc @arturkow2000

### CODE REVIEW

**Comment by daringer** [comment]
lgtm, this is something we should verify on actual hardware, just to be sure and as finding this as a source for issues at a later point might be painful

**Comment by robin-nitrokey** [comment]
Note that this is effectively the same patch that we are already using in nitrokey-3-firmware.

**Comment by nickray** [comment]
I'd be +1 on making this a breaking change, and adding whatever other improvements you can think of (e.g. your "should also change").

**Comment by robin-nitrokey** [comment]
Updated:
- Rebased onto 7b6685771fe350b8c532c04d4264e7a6d65d30c3.
- Replaced `LOOKAHEADWORDS_SIZE` (measured in 4 bytes) with `LOOKAHEAD_SIZE` (measured in 8 bytes) so that all possible values are valid.
- Bumped Rust version back to 1.67.0.

Not tested on hardware yet, also no unit test for the issue yet.  But at least the 1.67.0 segfault fix shows us that we are doing something right. :-)

I’d like to merge and release this next week as this a rather severe issue.

@trussed-dev/nitrokey Please review, test and comment.
@daringer Can you test this on the NK3AM with your load test script?
@arturkow2000 @lexxvir @svenstaro If you’re still working with `littlefs2`, would you mind testing this change in your projects?

File: src/fs.rs [hunk 1]
```diff
@@ -89,6 +87,10 @@ impl<Storage: driver::Storage> Allocation<Storage> {
         debug_assert!(cache_size <= block_size);
         debug_assert!(block_size % cache_size == 0);
 
+        // lookahead words size (measured in 4 bytes) must be a multiple of 2 so that the actual
+        // lookahead size is a multiple of 8 bytes
+        debug_assert!(lookahead_size % 2 == 0);
```
Comment (robin-nitrokey) [inline src/fs.rs:92]: Forgot about this one, has to be removed as we now use multiples of 8 bytes.

**Comment by nickray** [comment]
I'm wondering if we should have all the sizes in bytes, and use compile time checks to enforce multiples, and compile time additions for the "plus one" stuff -- if possible -- as it would make for a cleaner API.

**Comment by robin-nitrokey** [comment]
I don’t think this is possible at the moment because of the limitations with the const generics in stable.  We cannot use generic parameters in const operations, so we still need `generic-array` which cannot do arithmetics on the array size AFAIK.  And we cannot use generic parameters in const assertions.

I thought about adding an optional verification method that could be used to validate a `Storage` implementation, but it is nothing we can enforce.  So I’d prefer to stick with the current approach for compile-time checks.

**Comment by sosthene-nitrokey** [comment]
> And we cannot use generic parameters in const assertions.

This can be done by having an associated constant for the const assertion: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=40f243447104f266f22aa71b3f18c359

**Comment by robin-nitrokey** [comment]
In this case, it is about accessing an associated constant of a generic parameter.  Moving all associated constant into parameters is unrealistic in this case.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=4e53f8caa18adb02be25667bbf27d3bc

**Comment by robin-nitrokey** [comment]
I think the tracking issue for this would be: https://github.com/rust-lang/rust/issues/76560

**Comment by daringer** [comment]
ok tested on nrf extensively and not on lpc55, but all adaptations needed are available as PRs:
* https://github.com/Nitrokey/nitrokey-3-firmware/pull/168 :arrow_backward: nrf adaptations & patches upgraded (also for lpc)
* https://github.com/Nitrokey/lpc55-hal/pull/1 :arrow_backward: lpc adaptations

`nk3xn` & `nk3am.bl` compile and `nk3am.bl` has seen ~200 RK operations, ~15 resets, ~50 pin-sets/changes w/o any issues. 

Soooo, lgtm :dancers:

**Comment by robin-nitrokey** [comment]
Nice!  Thanks @daringer!  I’ve tested it briefly on lpc55 with the NK3CN (provisioning FIDO2, setting a PIN, registration and authentication with an resident key) and everything worked.  For me this PR is ready to merge – any objections?

**Comment by daringer** [comment]
nope, let's go

**Comment by nickray** [comment]
Before we do so, please remove `rust-toolchain.toml` instead of bumping it.
I have a fix but couldn't push on top of this PR.

**Comment by sosthene-nitrokey** [comment]
Regarding the question on whether this is a breaking change or not, it doesn't really matter as  5d31dc5ebdfeafcbf50d9b4f95e888b98013b79a is clearly a breaking change, so the next release will need to be a major semver bump.

**Comment by szszszsz** [comment]
Suggestion: test all applications next time, not only fido-auth.

**Pull request merged**